common/src/main/java/org/conscrypt/ConscryptFileDescriptorSocket.java

```diff
@@ -271,8 +271,14 @@ public final void startHandshake() throws IOException {
 
             // Restore the original timeout now that the handshake is complete
             if (handshakeTimeoutMilliseconds >= 0) {
-                setSoTimeout(savedReadTimeoutMilliseconds);
-                setSoWriteTimeout(savedWriteTimeoutMilliseconds);
+                try {
+                    setSoTimeout(savedReadTimeoutMilliseconds);
+                    setSoWriteTimeout(savedWriteTimeoutMilliseconds);
+                } catch (SocketException e) {
+                    SSLHandshakeException wrapper =  new SSLHandshakeException(e.getMessage());
+                    wrapper.initCause(e);
+                    throw wrapper;
```

Comment:
Actually this might need more work.  If we get here we probably need to release resources below.